Add fastRepaintRectInLocalCoordinates

Constellation opened the pull request:
#### fa7447aa7e693cc4500f00fe4aa55de4d5dcf993
<pre>
Add fastRepaintRectInLocalCoordinates
<a href="https://bugs.webkit.org/show_bug.cgi?id=262667">https://bugs.webkit.org/show_bug.cgi?id=262667</a>
rdar://116492451

Reviewed by Said Abou-Hallawa.

Add fastRepaintRectInLocalCoordinates. We agreed that we will keep
repaintRectInLocalCoordinates as is. And repaintRectInLocalCoordinates will
be used only for RenderTreeAsText and checkEnclosure related things.
So approximate one will get fastRepaintRectInLocalCoordinates and
repainting code will use fastRepaintRectInLocalCoordinates.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::fastRepaintRectInLocalCoordinates const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.h:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/RenderSVGGradientStop.h:
* Source/WebCore/rendering/svg/RenderSVGImage.h:
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/RenderSVGText.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:

Canonical link: <a href="https://commits.webkit.org/268889@main">https://commits.webkit.org/268889@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f18d319ac72494541a46bbfb56f01e0038ceb804

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20985 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22866 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24618 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21551 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21208 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20953 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/18209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23720 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/18113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19033 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/19198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/19234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/23235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/19780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/19040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2590 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19615 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -134,6 +134,7 @@ class RenderBox : public RenderBoxModelObject {
     void addFocusRingRects(Vector<LayoutRect>&, const LayoutPoint& additionalOffset, const RenderLayerModelObject* paintContainer = nullptr) const override;
     
     FloatRect repaintRectInLocalCoordinates() const override { return borderBoxRect(); }
+    FloatRect approximateRepaintRectInLocalCoordinates() const override { return borderBoxRect(); }
```
Not calling `repaintRectInLocalCoordinates` to avoid virtual function dispatch twice.

shallawa commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -507,6 +507,10 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
 
+    // Returns approximate rectangle enclosing all of the painted content.
+    // This is always larger-or-equal to repaintRectInLocalCoordinates().
+    virtual FloatRect approximateRepaintRectInLocalCoordinates() const;
```
Can we rename this `fastRepaintRectInLocalCoordinates()` to be consistent with `Path::boundingRect()` and `Path::fastBoundingRect()`?

Constellation replied:
Sounds good, done.

alanbaradlay replied:
if we end up polluting non-svg renderers with SVG specific behavior, please make it clear in the comment.

Constellation replied:
OK, added a comment here.

Constellation committed:
Add fastRepaintRectInLocalCoordinates
https://bugs.webkit.org/show_bug.cgi?id=262667
rdar://116492451

Reviewed by Said Abou-Hallawa.

Add fastRepaintRectInLocalCoordinates. We agreed that we will keep
repaintRectInLocalCoordinates as is. And repaintRectInLocalCoordinates will
be used only for RenderTreeAsText and checkEnclosure related things.
So approximate one will get fastRepaintRectInLocalCoordinates and
repainting code will use fastRepaintRectInLocalCoordinates.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::fastRepaintRectInLocalCoordinates const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.h:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/RenderSVGGradientStop.h:
* Source/WebCore/rendering/svg/RenderSVGImage.h:
* Source/WebCore/rendering/svg/RenderSVGInline.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGShape.h:
* Source/WebCore/rendering/svg/RenderSVGText.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:

Canonical link: https://commits.webkit.org/268889@main
